include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_parallel_backend_sycl_scan_H
+#define _ONEDPL_parallel_backend_sycl_scan_H
+
+#include "../../pstl/hetero/dpcpp/sycl_defs.h"
+#include "../../pstl/hetero/dpcpp/unseq_backend_sycl.h"
+#include "../../pstl/hetero/dpcpp/parallel_backend_sycl.h"
+
+namespace oneapi::dpl::experimental::kt
+{
+
+namespace gpu
+{
+
+namespace __impl
+{
+
+template <typename... _Name>
+class __lookback_init_kernel;
+
+template <typename... _Name>
+class __lookback_kernel;
+
+static constexpr int SUBGROUP_SIZE = 32;
+
+template <typename _T>
+struct __scan_status_flag
+{
+    using _FlagType = uint32_t;
+    using _AtomicFlagT = sycl::atomic_ref<_FlagType, sycl::memory_order::acq_rel, sycl::memory_scope::device,
+                                          sycl::access::address_space::global_space>;
+    using _AtomicValueT = sycl::atomic_ref<_T, sycl::memory_order::acq_rel, sycl::memory_scope::device,
+                                           sycl::access::address_space::global_space>;
+
+    static constexpr _FlagType __initialized_status = 0;
+    static constexpr _FlagType __partial_status = 1;
+    static constexpr _FlagType __full_status = 2;
+    static constexpr _FlagType __oob_status = 3;
+
+    static constexpr int __padding = SUBGROUP_SIZE;
+
+    __scan_status_flag(_FlagType* __flags, _T* __full_vals, _T* __partial_vals, const std::uint32_t __tile_id)
+        : __tile_id(__tile_id), __flags_begin(__flags), __full_vals_begin(__full_vals),
+          __partial_vals_begin(__partial_vals), __atomic_flag(*(__flags + __tile_id + __padding)),
+          __atomic_partial_value(*(__partial_vals + __tile_id + __padding)),
+          __atomic_full_value(*(__full_vals + __tile_id + __padding))
+    {
+    }
+
+    void
+    set_partial(const _T __val)
+    {
+        __atomic_partial_value.store(__val, sycl::memory_order::release);
```

Comment:
Do we need a memfence operation between the partial value and flag stores to guarantee the operations are not reordered?

Otherwise, I think we may be at risk of flag being updated first which could lead to reading a partial value before one is present.

Reply:
In my understanding, if a consumer work-item observes the write to the flag, then that work-item will also observe the write to the partial value since the work-items are synchronizing on the same atomic variable using release-acquire. When the producer work-item releases the flag, all writes that occurred by the producer up to that point will be observed by the consumer work-item when it acquires the flag. Since the writes to the partial value and flag cannot be reordered within the work-item (as they are tagged with release semantics), then it shouldn't be possible for the consumer work-item to observe an old partial value if it observes an update to the flag.

At least, that is how I understood the SYCL memory consistency model (§3.8.3) and how it adapts the standard C++ memory consistency model.

Reply:
Thanks for the pointers. I have read through the semantics of acquire-release and agree that this is safe.

I no longer believe there is a risk for the operations to be reordered.